CBL-59: N1QL date/time functions

callumbirks opened the pull request:
Borrowed some work from Jim's branch `feature/date_part_functions`. Didn't rebase because it messed with a lot of things.
All functions in the Epic https://issues.couchbase.com/browse/CBL-59 are now available in N1QL. Tests are added in this PR for those functions, and if you inspect the tests you can compare them against the PRD which is linked in the Epic.
Per our conversation with PM, I've changed the `tz` functions to use an offset (in minutes) rather than a Timezone string.

callumbirks commented:
Ignore Blackduck failures for now, I am just testing the Fleece PR to make sure Core's CI passes with it.

borrrden commented on LiteCore/Query/SQLiteDateTimeHelpers.hh:
```diff
@@ -0,0 +1,220 @@
+        return (year / 4) - (year / 100) + (year / 400);
+    }
+
+    inline int64_t leapYearsBetween(const int64_t start, const int64_t end) {
```
I'm concerned about why this is necessary....doesn't date.h already take leap years into account with its math operators?

callumbirks commented:
Discussed with @borrrden, I am going to replace some of the hand-written functionality with alternatives from "date.h".

callumbirks commented:
Wow, not sure what happened with the commits here.....

callumbirks commented:
Idk what happened to my local git but somehow this branch had commits from master, but wasn't based on an up-to-date master, so had a bunch of extra commits.
Rebased and squashed to fix.

snej commented on LiteCore/Query/SQLiteDateTimeHelpers.hh:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    inline bool parseDateArgRaw(sqlite3_value* arg, DateTime* outTime) {
+        if ( sqlite3_value_type(arg) != SQLITE_TEXT ) return false;
+        const auto str = valueAsStringSlice(arg);
```
You could replace both lines with a call to `stringSliceArgument`.

callumbirks replied:
Ahhh! I would just change it, but I don't want to wait another day for Jenkins failurs 😭
